File: sparktk-core/src/test/scala/org/trustedanalytics/sparktk/graph/internal/ops/BetweennessCentralityTest.scala
/**
 *  Copyright (c) 2016 Intel Corporation 
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package org.trustedanalytics.sparktk.graph.internal.ops

import org.scalatest.Matchers
import org.apache.spark.sql.{ SQLContext, Row }
import org.trustedanalytics.sparktk.testutils.TestingSparkContextWordSpec
import org.apache.spark.sql.catalyst.expressions.GenericRow
import org.trustedanalytics.sparktk.graph.Graph

class BetweennessCentralityTest extends TestingSparkContextWordSpec with Matchers {

  "Betweenness Centrality" should {
    def getGraph: Graph = {
      val sqlContext = new SQLContext(sparkContext)
      // Creates an unlabeled petersen graph
      val vertices = sqlContext.createDataFrame(Array(
        (0L, "a"),
        (1L, "b"),
        (2L, "c"),
        (3L, "d"),
        (4L, "e"),
        (5L, "f"),
        (6L, "g"),
        (7L, "h"),
        (8L, "i"),
        (9L, "j"))).toDF("id", "value")
      // create routes RDD with srcid, destid, distance
      val edges = sqlContext.createDataFrame(Array(
        (1L, 3L, 1.0),
        (1L, 4L, 1.0),
        (2L, 5L, 1.0),
        (3L, 5L, 1.0),
        (4L, 2L, 1.0),

        (0L, 2L, 1.0),
        (1L, 6L, 1.0),
        (5L, 7L, 1.0),
        (4L, 8L, 1.0),
        (3L, 9L, 1.0),

        (6L, 0L, 1.0),
        (0L, 9L, 1.0),
        (9L, 8L, 1.0),
        (8L, 7L, 1.0),
        (7L, 6L, 1.0))).toDF("src", "dst", "weight")
      // define the graph
      new Graph(vertices, edges)
    }

    "calculate betweenness centrality on the petersen graph in sparktk" in {
      val betweennessGraph = getGraph.betweennessCentrality()
      betweennessGraph.collect().toArray.toList should contain theSameElementsAs (
        List(new GenericRow(Array(4, "e", 0.08333333333333333)),
          new GenericRow(Array(0, "a", 0.08333333333333333)),
          new GenericRow(Array(1, "b", 0.08333333333333333)),
          new GenericRow(Array(6, "g", 0.08333333333333333)),
          new GenericRow(Array(3, "d", 0.08333333333333333)),
          new GenericRow(Array(7, "h", 0.08333333333333333)),
          new GenericRow(Array(9, "j", 0.08333333333333333)),
          new GenericRow(Array(8, "i", 0.08333333333333333)),
          new GenericRow(Array(5, "f", 0.08333333333333333)),
          new GenericRow(Array(2, "c", 0.08333333333333333))))
    }

    "calculate betweenness centrality on a 3 vertex line in sparktk" in {
      val sqlContext = new SQLContext(sparkContext)
      val vertices = sqlContext.createDataFrame(Array(
        (1L, "d"),
        (2L, "e"),
        (3L, "j"))).toDF("id", "value")
      // create routes RDD with srcid, destid, distance
      val edges = sqlContext.createDataFrame(Array(
        (1L, 2L, 1.0),
        (2L, 3L, 1.0)
      )).toDF("src", "dst", "weight")
      // define the graph
      val graph = new Graph(vertices, edges)
      val betweennessGraph = graph.betweennessCentrality()
      betweennessGraph.collect().toArray.toList should contain theSameElementsAs (
        List(new GenericRow(Array(1, "d", 0.0)),
          new GenericRow(Array(2, "e", 1.0)),
          new GenericRow(Array(3, "j", 0.0))))
    }

    "calculate betweenness centrality on a 6 vertex grid (ladder graph) in sparktk" in {
      val sqlContext = new SQLContext(sparkContext)
      val vertices = sqlContext.createDataFrame(Array(
        (0L, "d"),
        (1L, "e"),
        (2L, "f"),
        (3L, "g"),
        (4L, "h"),
        (5L, "i"))).toDF("id", "value")

      // create routes RDD with srcid, destid, distance
      val edges = sqlContext.createDataFrame(Array(
        (0L, 1L, 1.0),
        (0L, 2L, 1.0),
        (1L, 3L, 1.0),
        (2L, 3L, 1.0),
        (4L, 5L, 1.0),
        (2L, 4L, 1.0),
        (3L, 5L, 1.0)
      )).toDF("src", "dst", "weight")
      // define the graph
      val graph = new Graph(vertices, edges)
      val betweennessGraph = graph.betweennessCentrality(normalize = false)
      betweennessGraph.collect().toArray.toList should contain theSameElementsAs (
        List(new GenericRow(Array(4, "h", 0.8333333432674408D)),
          new GenericRow(Array(0, "d", 0.8333333432674408D)),
          new GenericRow(Array(1, "e", 0.8333333432674408D)),
          new GenericRow(Array(5, "i", 0.8333333432674408D)),
          new GenericRow(Array(3, "g", 3.3333333730697634D)),
          new GenericRow(Array(2, "f", 3.3333333730697634D))))

    }

    "calculate weighted betweenness centrality in sparktk" in {
      val sqlContext = new SQLContext(sparkContext)
      val vertices = sqlContext.createDataFrame(Array(
        (0L, "a"),
        (1L, "d"),
        (2L, "e"),
        (3L, "j"),
        (4L, "k"),
        (5L, "l"))).toDF("id", "value")
      // create routes RDD with srcid, destid, distance
      val edges = sqlContext.createDataFrame(Array(
        (0L, 1L, 3),
        (0L, 2L, 2),
        (0L, 3L, 6),
        (0L, 4L, 4),
        (1L, 3L, 5),
        (1L, 5L, 5),
        (2L, 4L, 1),
        (3L, 4L, 2),
        (3L, 5L, 1),
        (4L, 5L, 4)
      )).toDF("src", "dst", "weight")
      // define the graph
      val graph = new Graph(vertices, edges)
      val betweennessGraph = graph.betweennessCentrality(Some("weight"), normalize = false)
      betweennessGraph.collect().toArray.toList should contain theSameElementsAs (
        List(new GenericRow(Array(0, "a", 2.0)),
          new GenericRow(Array(1, "d", 0.0)),
          new GenericRow(Array(2, "e", 4.0)),
          new GenericRow(Array(3, "j", 3.0)),
          new GenericRow(Array(4, "k", 4.0)),
          new GenericRow(Array(5, "l", 0.0))))
    }
  }
}

File: regression-tests/sparktkregtests/testcases/graph/betweenness_centrality_test.py
# vim: set encoding=utf-8

#  Copyright (c) 2016 Intel Corporation 
#
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.
#

"""Tests betweenness centrality algorithm for graphs"""
import unittest
from sparktkregtests.lib import sparktk_test


class BetweennessCentrality(sparktk_test.SparkTKTestCase):

    def setUp(self):
        edges = self.context.frame.create(
            [(0, 1, 1),
             (0, 2, 1),
             (2, 3, 2),
             (2, 4, 4),
             (3, 4, 2),
             (3, 5, 4),
             (4, 5, 2),
             (4, 6, 1)],
            ["src", "dst", "weights"])

        vertices = self.context.frame.create(
            [[0], [1], [2], [3], [4], [5], [6]], ["id"])

        self.graph = self.context.graph.create(vertices, edges)

    def test_default(self):
        """Test default settings"""
        result_frame = self.graph.betweenness_centrality()
        result = result_frame.to_pandas()

        # validate centrality values
        expected_value = {
            0: 0.333,
            1: 0.0,
            2: 0.533,
            3: 0.1,
            4: 0.433,
            5: 0.0,
            6: 0.0}

        for i, row in result.iterrows():
            id = row['id']
            self.assertAlmostEqual(
                row["betweenness_centrality"],
                expected_value[id],
                delta=0.001)

    def test_weights_single_shortest_path(self):
        """Tests weighted betweenness when only one shortest path present"""
        edges = self.context.frame.create(
            [(0, 1, 3), (0, 2, 2),
             (0, 3, 6), (0, 4, 4),
             (1, 3, 5), (1, 5, 5),
             (2, 4, 1), (3, 4, 2),
             (3, 5, 1), (4, 5, 4)],
            ["src", "dst", "weights"])
        vertices = self.context.frame.create(
            [[0], [1], [2], [3], [4], [5]], ["id"])
        graph = self.context.graph.create(vertices, edges)

        # validate against values from networkx betweenness centrality
        result_frame = graph.betweenness_centrality("weights", False)
        result = result_frame.to_pandas()
        expected_values = {
            0: 2.0,
            1: 0.0,
            2: 4.0,
            3: 3.0,
            4: 4.0,
            5: 0.0}

        for i, row in result.iterrows():
            id = row['id']
            self.assertAlmostEqual(
                row["betweenness_centrality"],
                expected_values[id],
                delta=0.1)

    @unittest.skip("Bug:DPNG-14802")
    def test_weights(self):
        """Test betweenness with weighted cost"""
        result_frame = self.graph.betweenness_centrality("weights", False)

        # validate betweenness centrality values
        expected_values = [0.0, 5.0, 0.0, 0.0, 8.0, 5.0, 7.5]
        result = result_frame.to_pandas()
        for i, row in result.iterrows():
            self.assertAlmostEqual(
                row["betweenness_centrality"],
                expected_values[i],
                delta=0.1)

    def test_disconnected_edges(self):
        """Test betweenness on graph with disconnected edges"""
        edges = self.context.frame.create(
            [['a', 'b'], ['a', 'c'],
             ['c', 'd'], ['c', 'e'],
             ['f', 'g'], ['g', 'h']],
            ['src', 'dst'])

        vertices = self.context.frame.create(
            [['a'], ['b'], ['c'], ['d'], ['e'], ['f'], ['g'], ['h']], ['id'])

        graph = self.context.graph.create(vertices, edges)

        result_frame = graph.betweenness_centrality(normalize=False)

        # validate betweenness centrality values
        expected_values = {
            'a': 3.0,
            'b': 0.0, 'c': 5.0, 'd': 0.0,
            'e': 0.0, 'f': 0.0, 'g': 1.0, 'h': 0.0}

        result = result_frame.to_pandas()
        for i, row in result.iterrows():
            id = row['id']
            self.assertAlmostEqual(
                row["betweenness_centrality"],
                expected_values[id],
                delta=0.1)

    def test_normalize(self):
        """Test unnomallized betweenness crentrality"""
        result_frame = self.graph.betweenness_centrality(normalize=False)
        result = result_frame.to_pandas()

        # validate centrality values
        expected_values = {
            0: 5.0,
            1: 0.0,
            2: 8.0,
            3: 1.5,
            4: 6.5,
            5: 0.0,
            6: 0.0}

        for i, row in result.iterrows():
            id = row['id']
            self.assertAlmostEqual(
                row["betweenness_centrality"],
                expected_values[id],
                delta=0.1)

    def test_bad_weights_column_name(self):
        """Should throw exception when bad weights column name given"""
        with self.assertRaisesRegexp(
                Exception, "Field \"BAD\" does not exist"):
            self.graph.betweenness_centrality("BAD")


if __name__ == "__main__":
    unittest.main()
